chain-extensions/pallet-assets/src/lib.rs

```diff
@@ -0,0 +1,371 @@
+use codec::Encode;
+use frame_support::traits::fungibles::InspectMetadata;
+use frame_support::traits::tokens::fungibles::approvals::Inspect;
+use pallet_assets::WeightInfo;
+use pallet_contracts::chain_extension::{
+    ChainExtension, Environment, Ext, InitState, RetVal, SysConfig,
+};
+use sp_runtime::traits::StaticLookup;
+use sp_runtime::DispatchError;
+use sp_std::marker::PhantomData;
+use sp_std::vec::Vec;
+
+enum AssetsFunc {
+    Create,
+    Transfer,
+    Mint,
+    Burn,
+    BalanceOf,
+    TotalSupply,
+    Allowance,
+    ApproveTransfer,
+    CancelApproval,
+    TransferApproved,
+    SetMetadata,
+    MetadataName,
+    MetadataSymbol,
+    MetadataDecimals,
+}
+
+impl TryFrom<u16> for AssetsFunc {
+    type Error = DispatchError;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        match value {
+            1 => Ok(AssetsFunc::Create),
+            2 => Ok(AssetsFunc::Transfer),
+            3 => Ok(AssetsFunc::Mint),
+            4 => Ok(AssetsFunc::Burn),
+            5 => Ok(AssetsFunc::BalanceOf),
+            6 => Ok(AssetsFunc::TotalSupply),
+            7 => Ok(AssetsFunc::Allowance),
+            8 => Ok(AssetsFunc::ApproveTransfer),
+            9 => Ok(AssetsFunc::CancelApproval),
+            10 => Ok(AssetsFunc::TransferApproved),
+            11 => Ok(AssetsFunc::SetMetadata),
+            12 => Ok(AssetsFunc::MetadataName),
+            13 => Ok(AssetsFunc::MetadataSymbol),
+            14 => Ok(AssetsFunc::MetadataDecimals),
+            _ => Err(DispatchError::Other(
+                "PalletAssetsExtension: Unimplemented func_id",
+            )),
+        }
+    }
+}
+
+/// Pallet Assets chain extension.
+pub struct AssetsExtension<T, W>(PhantomData<(T, W)>);
+
+impl<T, W> Default for AssetsExtension<T, W> {
+    fn default() -> Self {
+        AssetsExtension(PhantomData)
+    }
+}
+
+impl<T, W> ChainExtension<T> for AssetsExtension<T, W>
+where
+    T: pallet_assets::Config + pallet_contracts::Config,
+    <<T as SysConfig>::Lookup as StaticLookup>::Source: From<<T as SysConfig>::AccountId>,
+    <T as SysConfig>::AccountId: From<[u8; 32]>,
+    W: weights::WeightInfo,
+{
+    fn call<E: Ext>(&mut self, env: Environment<E, InitState>) -> Result<RetVal, DispatchError>
+    where
+        E: Ext<T = T>,
+    {
+        let func_id = env.func_id().try_into()?;
+        let mut env = env.buf_in_buf_out();
+
+        match func_id {
+            AssetsFunc::Create => {
+                let (origin, id, admin, min_balance): (
+                    Origin,
+                    <T as pallet_assets::Config>::AssetId,
+                    T::AccountId,
+                    T::Balance,
+                ) = env.read_as()?;
+
+                let base_weight = <T as pallet_assets::Config>::WeightInfo::create();
+                env.charge_weight(base_weight)?;
+
+                let call_result = pallet_assets::Pallet::<T>::create(
+                    Origin::get_origin(&origin, env).into(),
```

Comment:
One more comment/question about this part that I forgot - what is the use case when chain-extension acts as the original contract caller? This seems very insecure since it gives the contract essentially rights to do whatever they want with callers assets.

User can _approve_ the smart contract to move some assets using the `approve_transfer` call but that should be done via standard call to `pallet-assets::approve_transfer`.